src/poly/dcrt/matrix.rs

```diff
@@ -351,184 +558,518 @@ impl PolyMatrix for DCRTPolyMatrix {
+            |(cur_block_row_idx, next_block_row_idx)| {
+                parallel_iter!(col_offsets.iter().tuple_windows().collect_vec()).for_each(
+                    |(cur_block_col_idx, next_block_col_idx)| {
+                        let mut self_block_polys = self.block_entries(
+                            *cur_block_row_idx..*next_block_row_idx,
+                            *cur_block_col_idx..*next_block_col_idx,
+                        );
+                        let block_row_len = next_block_row_idx - cur_block_row_idx;
+                        let block_col_len = next_block_col_idx - cur_block_col_idx;
+                        for i in 0..block_row_len {
+                            for j in 0..block_col_len {
+                                self_block_polys[i][j] = -self_block_polys[i][j].clone();
+                            }
+                        }
+                        // This is secure because the modified entries are not overlapped among
+                        // threads
+                        unsafe {
+                            new_matrix.replace_block_entries(
+                                *cur_block_row_idx..*next_block_row_idx,
+                                *cur_block_col_idx..*next_block_col_idx,
+                                self_block_polys,
+                            );
+                        }
+                    },
+                );
+            },
+        );
+        new_matrix
+    }
+}
 
-        let result: Vec<Vec<DCRTPoly>> = parallel_iter!(0..nrow)
-            .map(|i| parallel_iter!(0..ncol).map(|j| &self.inner[i][j] * rhs).collect())
-            .collect();
+impl DCRTPolyMatrix {
+    fn new_empty(params: &DCRTPolyParams, nrow: usize, ncol: usize) -> Self {
+        let dim = params.ring_dimension() as usize;
+        let log_q_bytes = params.modulus_bits().div_ceil(8);
+        let entry_size = dim * log_q_bytes;
+        let len = entry_size * nrow * ncol;
+        let file = tempfile().expect("failed to open file");
+        file.set_len(len as u64).expect("failed to set file length");
+        if BLOCK_SIZE.get().is_none() {
+            if let Ok(block_size_str) = env::var("BLOCK_SIZE") {
+                let block_size =
+                    block_size_str.parse::<usize>().expect("failed to parse BLOCK_SIZE");
+                BLOCK_SIZE.set(block_size).unwrap();
+            } else {
+                let system = System::new_all();
+                let mem_size = system.total_memory() * 2 / 3;
+                #[cfg(feature = "parallel")]
+                let num_threads = rayon::current_num_threads();
+                #[cfg(not(feature = "parallel"))]
+                let num_threads = 1;
+                let num_polys = mem_size as usize / num_threads / entry_size;
+                let block_size = (num_polys as f64).sqrt() as usize;
+                BLOCK_SIZE.set(block_size).unwrap();
```

Comment:
Err value: 9769 when BLOCK_SIZE not set

Reply:
9769 is the block_size you obtained, right?

Reply:
ya but before i haven't set BLOCK_SIZE as env variable